process-engine/process_engine_core
# ✨ CallActivity termination

**Changes:**

1. Add notifications for `onProcessError` event.
2. Implement `ProcessInstanceStateHandlingFacade` for handling all state-related queries for a ProcessInstance.
    - This facade is designed to reduce the complexity of the `ExecuteProcessService` and `ResumeProcessService`.
    - The services will no longer handle state changes by themselves, but instead only call the corresponding functions on the facade.
3. Move all facades into a subfolder named `facades`.
4. Use the updated message paths and event types from the contracts package.
5. Fix the handling of `ProcessTerminated` messages in the handlers for `CallActivities`.
6. Derive `ParallelJoinGateway` from the default `FlowNodeHandler` and add proper handling for `ProcessTerminated` events.
    - The handler must not leave the handling of `ProcessTerminated` messages to the `FlowNodeHandlerInterruptible` base class, because that handler always cleans up its subscriptions during the `afterExecute` hook.
    - But since each instance of the `ParallelJoinGatewayHandler` is executed multiple times, due to the fact that it must wait for all incoming branches before continuing execution, the subscription must be kept alive, until the last branch has finished execution.
7. Fix termination of Subprocesses.

PR: #260

## How can others test the changes?

This is easiest to test with the BPMN Studio:
- Create a Process with a CallActivity.
- Create a Subprocess for that CallActivity - ideally with a suspendable task, so you can follow its progress.
- Manually stop either the subprocess, or the parent process with the CallActivity.
- See that both processes will terminate correctly.

## PR-Checklist

Please check the boxes in this list after submitting your PR:
